content/300-guides/050-database/870-using-prisma-with-mongodb.mdx

```diff
@@ -0,0 +1,351 @@
+
+<Admonition type="info">
+
+When using MongoDB, be aware that the “coupling” between your schema and the database is purposefully designed to be less rigid than with with SQL databases; MongoDB will not enforce the schema, so you have to verify data integrity.
+
+</Admonition>
+
+These iterative tasks of updating the schema and the database can result in inconsistencies between your schema and the actual data in the database. Let’s look at one scenario where this can happen, and then examine several strategies for you and your team to consider for handling these inconsistencies.
+
+**Scenario**: you need to add a new field to your schema to accommodate a billing address that is separate from the currently stored shipping address. You currently have the following schema in your `schema.prisma` file:
+
+```prisma file=prisma/schema.prisma
+model Product {
+  id     String  @id @default(auto()) @map("_id") @db.ObjectId
+  name   String  @unique
+}
+
+model Order {
+  id              String   @id @default(auto()) @map("_id") @db.ObjectId
+  product         Product  @relation(fields: [productId], references: [id])
+  shippingAddress Address
+  productId       String   @db.ObjectId
+}
+
+type Address {
+  street String @id @default(auto()) @map("_id") @db.ObjectId
+  city   String
+  zip    String
+}
+```
+
+There are a number of strategies you could use for migrating this schema:
+
+- **"On-demand" updates**: with this strategy, you and your team have agreed that updates can be made to the schema as needed. However, in order to avoid migration failures due to inconsistencies between the data and schema, there is agreement in the team that any new fields added are explicitly defined as optional.
+
+  In our scenario above, you can go ahead and add the new `billingAddress` field to the `User` model in your Prisma schema, and define it as optional:
+
+  ```prisma file=prisma/schema.prisma highlight=10;add
+  model Product {
+    id     String  @id @default(auto()) @map("_id") @db.ObjectId
+    name   String  @unique
+  }
+
+  model Order {
+    id              String   @id @default(auto()) @map("_id") @db.ObjectId
+    product         Product  @relation(fields: [productId], references: [id])
+    shippingAddress Address
+    billingAddress  Address?
+    productId       String   @db.ObjectId
+  }
+
+  type Address {
+    street String @id @default(auto()) @map("_id") @db.ObjectId
+    city   String
+    zip    String
+  }
+  ```
+
+  Then regenerate your Prisma Client using the `npx prisma generate` command. Next, update your application to reflect the new field, and redeploy your app. The records in the database will be updated as the application's users begin to enter their billing address in the new field.
```

Comment:
This might require some explanation why this works:
- New field was added that is allowed to be null - so the existing data fits to the new schema anyway.

(Is this correct @dpetrick?)

Reply:
Wrote some words on this for @tanberry 

Here's the scenario:
1. I have an existing app with existing data in my MongoDB database and an existing Prisma Schema
2. A feature we're implementing needs us to add a new phoneNumber field to the User model

In SQL, this field must be added via add column. Usually added as an optional field to avoid needing to backfill all old phone numbers. SQL won't let you add a required phoneNumber when the previous records are null.

MongoDB is different. There is nothing enforcing the schema. It's up to you as the developer to match the schema with your data.

- If you make the phoneNumber required, nothing's going to stop you. However, when you query users that don't have the phone number, you're going to get a runtime error, because the Prisma Schema says it's required, but the data shows differently.
- If you make the phoneNumber optional, what prisma (and mongoose) do is allow you to query the old users, where the phoneNumber is just undefined (or null, not sure). But say you're updating that phoneNumber, when you write the data back, you've set the phoneNumber

That's what I mean by migrating On-Demand (a term I made up). There's no specific migration step. Instead, your migrating old users over time as your application runs. It turns out this is how most people deal with changing data in MongoDB.

Reply:
> If you make the phoneNumber required, nothing's going to stop you. However, when you query users that don't have the phone number, you're going to get a runtime error, because the Prisma Schema says it's required, but the data shows differently.

Because of this, I would only consider this strategy valid when the phoneNumber is optional. Otherwise, if you apply this strategy, you end up with an application that can not handle any existing data.